### ISSUE TITLE
double colon make rules for complex build processes

### ISSUE THREAD

**Issue by TheSlowGrowth** [VCVRack/Rack#610]
OS (if applicable): all
Version (or "dev" if compiling from source): all

I'm working on some plugins. What I found is that I wanted some additional steps to be taken during the build process, namely compiling some resource files with python scripts:

```
all: resources
resources:
	cd $(SELF_DIR)/resourcebuilding && python3 build.py
```
(Frankly - I'm not a makefile guru. This might be totally wrong, but so far it seems to work for me)

Now I would like to extend the "clean" rule to also clean the generated files
```
clean:
	rm <someFiles>
```
which doesn't work and throws this error message:

> ../../plugin.mk:30: warning: overriding commands for target 'clean'
> Makefile:32: warning: ignoring old commands for target 'clean'

I've read about a solution using double colon rules to make something like this possible.
Is this an option?

**Comment by TheSlowGrowth** [comment]
hmm, I see. It would work as a hacky solution. But what if I wanted to run a script during the clean process?

BTW, apparently for the `all:` target, it works - or maybe it just works the way I invoke the makefiles right now (which is `make all` in the plugin directory and `make all` in the VCVRack main directory)

**Comment by AndrewBelt** [comment]
I don't think it's hacky. It's the supported way to add objects to the Rack plugin build system.

Why would you want to run a script during `make clean`?

**Comment by TheSlowGrowth** [comment]
Okay, so I'm in the process of editing my makefile to fit this scheme. 
I'm not sure if I can't make it work because it intrinsically can't work or if I'm doing it wrong.
This is what I have right now:
```
# this is part of a file included from my main makefile
SELF_DIR := $(dir $(lastword $(MAKEFILE_LIST)))
# resource input and output directories (absolute)
RESIN_DIR_ABS := $(shell pwd)/res
RESOUT_DIR_ABS := $(shell pwd)/src/res

SOURCES += $(RESOUT_DIR_ABS)fonts.cpp
SOURCES += $(RESOUT_DIR_ABS)images.cpp

all: resources

# create resources
resources:
	cd $(SELF_DIR)/resourcebuilding && python3 -m build -i "$(RESIN_DIR_ABS)" -o "$(RESOUT_DIR_ABS)"
```
Basically it's one pyhton script, generating a set of source files from a folder of images and fonts.
If I understand your suggestion, I'd have to add a "filename" to OBJECTS and instead of this being a simple file rule, I then supply my own rule for it in which I invoke the python script. The cleaning is then done by sumply making sure the generated files are inside the build folder.
First, I'd like to avoid having to put those files in the build folder, simply because they are included everywhere in the code and it seems super dirty to include stuff from a temporary build folder.
Secondly, what I don't get is how I can handle multiple files beeing created by the same script (which handles the "do I need to update anything?" by itself)
This is my attempt right now:
```
# this is part of a file included from my main makefile
SELF_DIR := $(dir $(lastword $(MAKEFILE_LIST)))
# resource input and output directories (absolute)
RESIN_DIR_ABS := $(shell pwd)/res
RESOUT_DIR_ABS := $(shell pwd)/src/res

SOURCES += $(RESOUT_DIR_ABS)fonts.cpp
SOURCES += $(RESOUT_DIR_ABS)images.cpp

OBJECTS += $(RESOUT_DIR_ABS)fonts.h
OBJECTS += $(RESOUT_DIR_ABS)fonts.cpp
OBJECTS += $(RESOUT_DIR_ABS)images.h
OBJECTS += $(RESOUT_DIR_ABS)images.cpp

$(RESOUT_DIR_ABS)fonts.h: resources
$(RESOUT_DIR_ABS)fonts.cpp: resources
$(RESOUT_DIR_ABS)images.h: resources
$(RESOUT_DIR_ABS)images.cpp: resources

# create resources
resources:
	cd $(SELF_DIR)/resourcebuilding && python3 -m build -i "$(RESIN_DIR_ABS)" -o "$(RESOUT_DIR_ABS)"
```
This correctly invokes the script (just once, as desired and expected) but then fails with a bunch of:
> clang: warning: treating 'c-header' input as 'c++-header' when in C++ mode, this behavior is deprecated
> clang: error: cannot specify -o when generating multiple output files

What also puzzles me is that it makes the script and the makefile interdependent. Whenever I decide to have the script create more files (or the filenames change) I have to also add all these rules to the makefile.

Am I doing this wrong?

**Comment by AndrewBelt** [comment]
Oh, my mistake. You can't add non-.o files to OBJECTS because they're appended to the compiler arguments.

Try this https://stackoverflow.com/a/1645087/272642 Hope this helps.

**Comment by TheSlowGrowth** [comment]
The stackoverflow solution assumes I can edit the global makefile (by which I mean the one of VCVRack)
I've seen this too when I searched for a solution, and it seems all possible solutions require to add something to the global makefile. That's why I opened this issue.

IMO the cleanes solution would be to add double colons to the `all:` and `clean:` rules, and possibly some other rules to which plugins might want to append something.
